accepted/0000-yomi-initial-integration.md

```diff
@@ -0,0 +1,361 @@
+definition of:
+
+- Partitions and file systems used in the node
+
+- Bootloader parameters
+
+- Repositories and packages installed
+
+- Services enabled and disabled
+
+- Users and certificates
+
+As we can see, this proposal is (1) moving closer the features of
+Kickstart based systems into the SUSE family, and (2) using SaltStack
+to orchestrate the new installation.
+
+# Detailed design
+[design]: #detailed-design
+
+## Integrating Yomi as a Salt state
+
+The core of the proposal is to add Yomi into the catalog of formulas
+available in Uyuni.
+
+As Yomi is indeed a Salt formula, this integration can be done
+directly, installing the `yomi-formula` package into the Manager.
+
+The user can later decide to add the Yomi state into the `top.sls` for
+this system, fill the pillar data using the SaltForm provided in the
+package, and launch the installer by applying the highstate.
+
+## SaltForms and sub-forms
+
+Uyuni provides a mechanism, SaltForms, that allows the creation of the
+pillar data that suit the formula. The current relation between a
+SaltForm and a Salt state is 1:1, but using some empty `init.sls` we
+have the option of splitting a big SaltForm into multiple smaller
+SaltForms (subforms), that optionally belongs to the same group.
+
+The current `yomi-formula` is using this mechanism to split the pillar
+data into six sections:
+
+- yomi: general configuration data, like use `kexec` instead of
+  `reboot` after the installation, or some `systemd-firstboot` data.
+
+- yomi-bootloader: indicate the device where the bootloader will be
+  installed, and the parameters send to the kernel.
+
+- yomi-services: list of services that will be enabled, and is
+  `salt-minion` data will be copied into the new system.
+
+- yomi-software: repository registering and packages installation.
+
+- yomi-storage: device partition, LVM or RAID configuration, and file
+  systems applied.
+
+- yomi-users: creation of initial users, passwords and certificates.
+
+As we will see in future sections, those forms contains some
+overlapping with other salt-forms (the locale configuration that
+`systemd-firstboot` requires), and with other subsystems in Uyuni
+(like the software section).
+
+## Highstate and Workflow
+
+Installing a new system will be equivalent of applying a highstate to
+a system booted from the Yomi image.
+
+The current Yomi image is a JeOS based image with a `salt-minion`,
+some systemd overwrites that will help during the initial
+configuration, and the CLI tools required when a new installation is
+performed. You can find more information here:
+https://github.com/openSUSE/yomi#the-iso-image
+
+The current Yomi states are aware of differentiating when the system
+is booting from an external image (like the Yomi image), from when the
+boot comes from the local device. This is used to decide when the
+installation of the node makes sense or not when a highstate is
+re-applied.
```

Comment:
Saltboot has similar thing - when minion is running within saltboot inird, there is extra grain (`saltboot_initrd`) set to true. Is it done similarly for Yomi? If so, how about also changing the Uyuni registration handler to automatically apply yomi state to trigger the installation on yomi startup from initrd? (see [RegisterMinionEventMessageAction.java#L359](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java#L359) )

Reply:
Good question.

I try to make Yomi as much idempotent as possible, so I put in place a mechanism in the Yomi state that will detect from where it is booting. So, for example, if the root file system device (indicated in the pillar) is already mounted as the current root_fs, Yomi have nothing to do and no installation will be done during the highstate.

But if the roof_fs is not mounted and is available in the system, Yomi can apply the rest of the states.

So there is no grain to signalize if an installation must be done, as Yomi have all the information.

But interestingly enough I propose a new grain later, that will be use for Yomi to signal to the manager to not apply the highstare automatically.

Reply:
I'd always try to put logic onto the minion, not the master. If the Yomi state can figure things out, it should.